sharding/nodesSetup.go

```diff
@@ -139,35 +168,69 @@ func (ns *NodesSetup) processShardAssignment() {
 	}
 }
 
-func (ns *NodesSetup) createInitialNodesPubKeys() {
+func (ns *NodesSetup) createInitialNodesInfo() {
 	nrOfShardAndMeta := ns.nrOfShards
 	if ns.MetaChainActive {
 		nrOfShardAndMeta += 1
 	}
 
-	ns.allNodesPubKeys = make(map[uint32][]string, nrOfShardAndMeta)
+	ns.allNodesInfo = make(map[uint32][]*NodeInfo, nrOfShardAndMeta)
 	for _, in := range ns.InitialNodes {
-		if in.pubKey != nil {
-			ns.allNodesPubKeys[in.assignedShard] = append(ns.allNodesPubKeys[in.assignedShard], string(in.pubKey))
+		if in.pubKey != nil && in.address != nil {
+			ns.allNodesInfo[in.assignedShard] = append(ns.allNodesInfo[in.assignedShard],
+				&NodeInfo{in.assignedShard, in.pubKey, in.address})
 		}
 	}
 }
 
-// InitialNodesPubKeys - gets initial public keys
+// InitialNodesPubKeys - gets initial nodes public keys
 func (ns *NodesSetup) InitialNodesPubKeys() map[uint32][]string {
-	return ns.allNodesPubKeys
+	allNodesPubKeys := make(map[uint32][]string, 0)
+	for shardId, nodesInfo := range ns.allNodesInfo {
+		pubKeys := make([]string, 0)
```

Comment:
I think it would be more efficient if you allocate the right size and just add on the position instead of append, when you already know the size

Reply:
Done